mpqp/execution/providers/azure.py

```diff
@@ -0,0 +1,202 @@
+    job: Optional[Job],
+    device: AZUREDevice,
+) -> Result:
+    from qiskit.result import Result as QiskitResult
+
+    if isinstance(result, QiskitResult):
+        if job is None:
+            job_data = result.data()
+            if "statevector" in job_data:
+                job_type = JobType.STATE_VECTOR
+                nb_qubits = int(math.log(len(result.get_statevector()), 2))
+                job = Job(job_type, QCircuit(nb_qubits), device)
+            elif "counts" in job_data:
+                job_type = JobType.SAMPLE
+                nb_qubits = len(list(result.get_counts())[0])
+                shots = result.results[0].shots
+                job = Job(
+                    job_type,
+                    QCircuit(nb_qubits),
+                    device,
+                    BasisMeasure(list(range(nb_qubits)), shots=shots),
+                )
+            else:
+                if len(result.data()) == 0:
+                    raise ValueError(
+                        "Result data is empty, cannot extract anything. Check "
+                        "if the associated job was successfully completed."
+                    )
+                else:
+                    raise ValueError(
+                        f"Data with keys {result.data().keys()} in result not handled."
+                    )
+
+        if job.job_type == JobType.STATE_VECTOR:
+            vector = np.array(result.get_statevector())
+            state_vector = StateVector(
+                vector,  # pyright: ignore[reportArgumentType]
+                job.circuit.nb_qubits,
+            )
+            return Result(job, state_vector, 0, 0)
+        elif job.job_type == JobType.SAMPLE:
+            assert job.measure is not None
+            job_data = result.data()
+            data = [
+                Sample(
+                    bin_str="".join(map(str, state)),
+                    probability=(
+                        job_data.get("probabilities").get(state)
+                        if "probabilities" in job_data
+                        else None
+                    ),
+                    nb_qubits=job.circuit.nb_qubits,
+                    count=int(count),
+                )
+                for (state, count) in job_data.get("counts").items()
+            ]
+            return Result(job, data, None, job.measure.shots)
+        else:
+            raise NotImplementedError(f"{job.job_type} not handled.")
+    else:
+        if job is None:
+            job = Job(JobType.OBSERVABLE, QCircuit(1), device)
+        return Result(job, 0, result.data())
+
+
+@typechecked
+def get_result_from_azure_job_id(job_id: str) -> Result:
+    """Retrieves from Azure remote platform and parse the result of the job_id
+    given in parameter. If the job is still running, we wait (blocking) until it
+    is ``DONE``.
+
+    Args:
+        job_id: Id of the remote Azure job.
+
+    Returns:
+        The result converted to our format.
+    """
```

Comment:
I would have imagined for this method something in the lines of
```py
job = get_jobs_by_id(job_id)
result = job.get_results()
return extract_result(result)
```
why is this not correct ?

Reply:
I have started working with this, but the result from a job id is not the same as the result of a run; it is not a Qiskit result.

Reply:
in azure.py:
```py 
88 def get_result_from_azure_job_id:
101 	 result = job.get_results()
```
return :
```py
typing.Any
Results dictionary with histogram shots, or raw results if not a json object. 
-> (dict[Any, Any] | Any | dict[Unknown, Unknown] | Unknown)
```

so if we pass it to extract_result
we need to pass result to `Any`

we can do it but we will lost the type_check

Reply:
And if I put the extraction code in `extract_resul`, there's no need to create `get_result_from_azure_job_id`.

Reply:
> And if I put the extraction code in `extract_result`, there's no need to create `get_result_from_azure_job_id` except for passing the job_id.

About this comment, even if we put this extraction here in `extract_result` the method of getting the result from azure job id is still useful, because it is the one that the user will use directly by just inputting the id. In our building of the execution process, and user never calls by himself the method extract_result, that is why I prefer to move everything there.

Reply:
> in azure.py:
> 
> ```python
> 88 def get_result_from_azure_job_id:
> 101 	 result = job.get_results()
> ```
> 
> return :
> 
> ```python
> typing.Any
> Results dictionary with histogram shots, or raw results if not a json object. 
> -> (dict[Any, Any] | Any | dict[Unknown, Unknown] | Unknown)
> ```
> 
> so if we pass it to extract_result we need to pass result to `Any`
> 
> we can do it but we will lost the type_check

Ummm this is a bit annoying that the providers return `Any` but we can't really do nothing on that part.

However, a possible escape is to define a function `extract_remote_result` or something like this specifically for this case, move it to `azure.py`, and then call this function from here. What do you think?

Reply:
Ok but if it is juste for `get_result_from_azure_job_id`, Is it still useful to make a second function if it's only used ones?